gpii/node_modules/userListeners/src/userFolder.js

```diff
@@ -0,0 +1,115 @@
+
+// The user folder user listener.
+fluid.defaults("gpii.userListeners.userFolder", {
+    gradeNames: ["fluid.component", "fluid.contextAware", "gpii.userListener"],
+    contextAwareness: {
+        platform: {
+            checks: {
+                windows: {
+                    contextValue: "{gpii.contexts.windows}",
+                    gradeNames: "gpii.userListeners.userFolder.windows"
+                }
+            }
+        }
+    },
+    members: {
+        tokenDirectory: "@expand:{settingsDir}.getGpiiSettingsDir()",
+        tokenFile: ".gpii-user-token.txt",
+        done: false,
+        listenerName: "userFolder"
+    },
+    components: {
+        settingsDir: {
+            type: "gpii.settingsDir"
+        }
+    },
+    invokers: {
+        startListener: "{that}.attemptLogin",
+        stopListener: "fluid.identity",
+        attemptLogin: {
+            funcName: "gpii.userListeners.userFolder.attemptLogin",
+            args: ["{that}"]
+        },
+        getToken: {
+            func: "{that}.readTokenFile",
+            args: ["{that}.tokenDirectory", "{that}.tokenFile"]
+        },
+        storeToken: {
+            func: "{that}.writeTokenFile",
+            args: [ "{that}.tokenDirectory", "{that}.tokenFile", "{arguments}.0" ]
+        }
+    },
+    listeners: {
+        "onCreate.loginListener": {
+            funcName: "gpii.userListeners.userFolder.addLoginListener",
+            args: [ "{that}", "{lifecycleManager}.events.onSessionStart"]
+        }
+    },
+    // Disabling by default, as it is the safest and will prevent people's keys from being accidentally stored when it
+    // is not desired (eg, on shared or public access computers).
+    saveLastLogin: false
+});
+
+/**
+ * Adds a listener to the onSessionStart event of the lifecycle manager, which stores the token file.
+ *
+ * @param {Component} that The gpii.userListeners.userFolder instance.
+ * @param {EventFirer} firer The event firer for the onSessionStart event.
+ */
+gpii.userListeners.userFolder.addLoginListener = function (that, firer) {
+    if (that.options.saveLastLogin) {
+        firer.addListener(function (gradeName, gpiiKey) {
+            if (gpiiKey !== "noUser") {
+                that.storeToken(gpiiKey);
+            }
+        });
+    }
+};
+
+/**
+ * Performs the key-in, by raising the onTokenArrive event with the result of getToken().
+ *
+ * This should only be called once, during startup. Subsequent calls will have no effect.
+ *
+ * @param {Component} that - The gpii.userListeners.userFolder instance.
+ * @return {Promise} A promise resolving when the token has been read.
+ */
+gpii.userListeners.userFolder.attemptLogin = function (that) {
+    var promise;
+
+    if (that.done) {
+        promise = fluid.promise().reject({
+            isError: true,
+            message: "userFolder.attemptLogin had already been called."
+        });
+    } else {
+        that.done = true;
```

Comment:
Shall this line be inside the `promise.then()` block down below so that in the case `that.getToken()` fails, `that.done` will not be set to `true`.